plugins/chat/app/services/chat/create_message_interaction.rb

```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+module Chat
+  # Service responsible for creating and validating a new interaction between a user and a message.
+  #
+  # @example
+  #  Chat::CreateMessageInteraction.call(params: { message_id: 3, action_id: "xxx" }, guardian: guardian)
+  #
+  class CreateMessageInteraction
+    include ::Service::Base
+
+    # @!method self.call(guardian:, params:)
+    #   @param [Guardian] guardian
+    #   @param [Hash] params
+    #   @option params [Integer] :message_id
+    #   @option params [Integer] :action_id
+    #   @return [Service::Base::Context]
+    params do
+      attribute :message_id, :integer
+      validates :message_id, presence: true
+
+      attribute :action_id, :string
+      validates :action_id, presence: true
+    end
+
+    model :message
+    policy :can_interact_with_message
+    model :action
+
+    transaction do
+      model :interaction
+      step :trigger_interaction
+    end
+
+    private
+
+    def fetch_message(params:)
+      Chat::Message.find_by(id: params.message_id)
+    end
+
+    def fetch_action(params:, message:)
+      message.blocks&.find do |item|
+        item["elements"].find { |element| element["action_id"] == params.action_id }
+      end
+    end
+
+    def can_interact_with_message(guardian:, message:)
+      guardian.can_preview_chat_channel?(message.chat_channel)
+    end
+
+    def fetch_interaction(guardian:, message:, action:)
```

Comment:
You are missing handling in the controller for this model not being found (AKA the create failing with errors). @Flink  and I were recently chatting about having nice ways for services to fail and expose model errors on create/update calls, until then in this PR I am doing this and using `step` instead of `model`:

```ruby
def save_theme(theme:)
  theme.save

  if theme.errors.any?
    fail!("Could not save theme with errors #{theme.errors.full_messages.join(",")}")
  end
end
```

So TLDR not sure if you want to handle this and add a controller `on_X` block as well.

Reply:
`on_model_not_found` won’t work here as `#create` will always return a model. If that model can have errors, however, you can use `on_model_errors` instead.
We could also rename that step from `fetch_interaction` to `create_interaction`. (`model :interaction, :create_interaction`)

@martin-brennan your example is slightly different because you have to call `#save`, so as of now we don’t have a mechanism to handle it automatically. But if you’re calling a method that returns a model, then the `model` step can be used as it will check if the model isn’t empty and if present, if it’s valid.

In the case the model has no validations, it also means the creation step can’t fail (unless from low-level errors, but that’s another story).

Reply:
How about I check if the model is valid and return nil if it's not?

Reply:
I don’t think there’s anything to do, except for renaming `fetch_interaction` to `create_interaction` as it’s more consistent. It was more an explanation than anything else :sweat_smile: